examples/diffusion/diffusion-cli.cpp

```diff
@@ -5,65 +5,192 @@
 #include "log.h"
 
 #include <limits.h>
-#include <string>
-#include <vector>
+
 #include <algorithm>
 #include <cmath>
+#include <cstring>
 #include <limits>
 #include <random>
+#include <string>
+#include <vector>
 
-typedef bool (*diffusion_step_callback_t)(int32_t step,
-                                          int32_t total_steps,
-                                          const llama_token * tokens,
-                                          int32_t n_tokens,
-                                          void * user_data);
-
-enum diffusion_alg {
-    DIFFUSION_ALG_ORIGIN       = 0,
-    DIFFUSION_ALG_MASKGIT_PLUS = 1,
-    DIFFUSION_ALG_TOPK_MARGIN  = 2,
-    DIFFUSION_ALG_ENTROPY      = 3,
+// Dream remasking algorithms
+enum diffusion_algorithm_dream {
+    ORIGIN       = 0,
+    MASKGIT_PLUS = 1,
+    TOPK_MARGIN  = 2,
+    ENTROPY      = 3,
+};
+
+// LLaDA remasking types
+enum diffusion_algorithm_llada {
+    LOW_CONFIDENCE = 0,
+    RANDOM         = 1,
 };
```

Comment:
Is this separation necessary? For example, can we use "RANDOM" sampling with Dream?